helper/genericdrive.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+ * Fedora Media Writer
+ * Copyright (C) 2016 Martin Bříza <mbriza@redhat.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include "genericdrive.h"
+
+#include <string>
+#include <stdexcept>
+
+#include <QString>
+#include <QTextStream>
+#include <QFile>
+#include <QFileInfo>
+
+#include "isomd5/libcheckisomd5.h"
+#include <lzma.h>
+
+#include "page_aligned_buffer.h"
+#include "write.h"
+
+#ifndef MEDIAWRITER_LZMA_LIMIT
+// 256MB memory limit for the decompressor
+#define MEDIAWRITER_LZMA_LIMIT (1024 * 1024 * 256)
+#endif
+
+void GenericDrive::writeCompressed(const QString &source) {
+    qint64 totalRead = 0;
+
+    lzma_stream strm = LZMA_STREAM_INIT;
+    lzma_ret ret;
+
+    PageAlignedBuffer<2> buffers;
+    const std::size_t bufferSize = buffers.size;
+    char *inBuffer = static_cast<char *>(buffers.get(0));
+    char *outBuffer = static_cast<char *>(buffers.get(1));
+    auto total = QFileInfo(source).size();
+    ProgressStats progress;
+    progress.fd = getDescriptor();
+
+    QFile file(source);
+    file.open(QIODevice::ReadOnly);
+
+    ret = lzma_stream_decoder(&strm, MEDIAWRITER_LZMA_LIMIT, LZMA_CONCATENATED);
+    if (ret != LZMA_OK) {
+        throw std::runtime_error("Failed to start decompressing.");
```

Comment:
Yes exceptions are not nice in Qt but should be safe in this context. Translations can be done for error messages but that's hard when non-predictable detail is provided in the error message that is separated from the translation. The question also is [if these exceptions should be translated](https://stackoverflow.com/q/2506959).
Note the current way of error handling also has hard to fix issues like throwing three error messages even though only instead of aborting after the first time see #96. Not using exceptions here also makes the code look bad IMHO since you have to pass the error down the chain which is what I mainly tried to avoid.

Reply:
You can use QString::arg to mark which parts are variable using the `%1, %2, ...` notation inside the string. These error messages would typically be displayed to the user. It's also possible to translate them on the fly and forward the untranslated string to the debug log, which would then be what the bug reporter would attach to the issue when reporting a bug.
You're right though the three-error-messages bug but I'd consider that rather a bad design choice than a bug. Overall, I'd say the helpers have a slight problem with error reporting - there are too many messages and they don't usually describe what the problem is to the user. That's not a subject of this PR though so I'd leave that be for now.

Reply:
Let's see what I can come up with.